pysisyphus/calculators/CFOUR.py

```diff
@@ -0,0 +1,191 @@
+        try:
+            self.initden = kept_fns["density"]
+        except KeyError:
+            self.log("den.dat not found!")
+            return
+
+    def prepare_input(self, atoms, coords, calc_type):
+        xyz_string = self.prepare_coords(atoms, coords, angstrom=False)
+        cfour_keyword_string = '\n'.join(f"{key.upper()}={str(value).upper()}" for key, value in self.cfour_input.items())
+        grad_string = "DERIV_LEVEL=1\n" if calc_type == "grad" else ""
+
+        ### Note: CFOUR abhors blank lines between keywords. Make sure no extra whitespace is added between keyword lines.
+        ### First dedent, then format.
+        input_str = textwrap.dedent("""\
+        CFOUR calculation
+        {xyz_string}
+
+        *CFOUR(UNITS=BOHR
+        COORDINATES=CARTESIAN
+        {grad_string}{cfour_keyword_string})
+
+        """).format(xyz_string=xyz_string, grad_string=grad_string, cfour_keyword_string=cfour_keyword_string)
+
+        return input_str
+
+    def get_energy(self, atoms, coords):
+        return self.run_calculation(atoms, coords, "energy")
+
+    def get_forces(self, atoms, coords):
+        return self.run_calculation(atoms, coords, "grad")
+
+    def parse_energy(self, path):
+        energy_fn = path / self.out_fn
+        with open(energy_fn) as handle:
+            text = handle.read()
+
+        regex = "\s*The final electronic energy is\s*" + self.float_regex
+        mobj = re.search(regex, text, re.DOTALL)
+        energy = float(mobj.groups()[0])
+
+        return {
+            "energy": energy
+        }
+
+    def parse_gradient(self, path):
+        ## Adapted from OpenMolcas calculator
+        results = {}
+        gradient_fn = path / self.gradient_fn
+        gradient = np.loadtxt(gradient_fn, skiprows=1)
+        natoms = int(gradient.shape[0] / 2)
+        gradient = gradient[natoms:, 1:]
+        with open(path / self.out_fn) as handle:
+            text = handle.read()
+
+        gradient_rotated = self.rotate_gradient(text, gradient).flatten()
+
+        energy = self.parse_energy(path)["energy"]
+        results["energy"] = energy
+        results["forces"] = -gradient_rotated
+
+        return results
+
+    def rotate_gradient(self, text, gradient):
+        ## Following http://nghiaho.com/?page_id=671
+        ## Permalink: https://web.archive.org/web/20230128004529/http://nghiaho.com/?page_id=671
+
+        cfour_coords_3d = self.read_geom(text)
+        pysis_coords_3d = np.reshape(self.input_coords, (-1,3))
+        cfour_centroid = self.calc_centroid(cfour_coords_3d)
+        pysis_centroid = self.calc_centroid(pysis_coords_3d)
+
+        rot_matrix = self.calc_rot_matrix(cfour_coords_3d, pysis_coords_3d, cfour_centroid, pysis_centroid)
+
+        return (gradient @ rot_matrix.T)
+
+    def read_geom(self, text):
+        regex = r"Coordinates used in calculation \(QCOMP\)(.+)Interatomic distance matrix"
+        floats =  [self.float_regex for i in range(3)]
```

Comment:
You can keep it like this. Personally I would have written `[self.float_regex] * 3`